### ISSUE TITLE
min_sum 3x slower on torch.compile (torch 2.0.1, CPU) for small 1d vectors (static shapes). dynamic shapes are quite unintuitive

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#106466]
### 🐛 Describe the bug

I would assume it should be faster because it should not need to allocate an intermediate float tensor and it could use the warmed-up small static shapes and produce unrolled loops.

`TORCH_LOGS=output_code python3 foo.py` dumps no logs/code (cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @cpuhrsch @albanD )

`TORCH_COMPILE_DEBUG=1 python3 foo.py` does work. It would be good if it also produced a zipball of the log directory for attachment (I did it manually with `zip -r torch_compile_debug.zip ./torch_compile_debug/` [torch_compile_debug.zip](https://github.com/pytorch/pytorch/files/12242115/torch_compile_debug.zip) ) according to https://pytorch.org/tutorials/intermediate/inductor_debug_cpu.html

```python
# foo.py
import torch
import time

min_sum = lambda a, b: torch.min(a, b).sum(-1)
min_sum_compiled = torch.compile(min_sum)

static_shapes = [
    ((192,), (192,)),
    ((96,), (96,))
]

# warmup
for k in range(100):
    for a_shape, b_shape in static_shapes:
        a = torch.rand(*a_shape, dtype = torch.float32)
        b = torch.rand(*b_shape, dtype = torch.float32)
        min_sum(a, b)
        min_sum_compiled(a, b)

for a_shape, b_shape in static_shapes:
    a = torch.rand(*a_shape, dtype = torch.float32)
    b = torch.rand(*b_shape, dtype = torch.float32)
    tic = time.time()
    for k in range(100):
        min_sum(a, b)
    print('min_sum', a_shape, b_shape, (time.time() - tic) * 1000, 'ms')

    tic = time.time()
    for k in range(100):
        min_sum(a, b)
    print('min_sum_compiled', a_shape, b_shape, (time.time() - tic) * 1000, 'ms')
```
```
min_sum (192,) (192,) 1.3213157653808594 ms
min_sum_compiled (192,) (192,) 4.303216934204102 ms
min_sum (96,) (96,) 1.3997554779052734 ms
min_sum_compiled (96,) (96,) 4.553318023681641 ms
```

I got two kernels. Should it include a pragma unroll? Why is it not vectorized? Will it be auto-vectorized by the compiler? How to find out if it was actually vectorized? If I don't care about NaNs, can I tell inductor to not include the guard ternary?

```cpp
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       const float* __restrict__ in_ptr1,
                       float* __restrict__ out_ptr0)
{
    {
        {
            float tmp3 = 0;
            for(long i0=0; i0<192; i0+=1)
            {
                auto tmp0 = in_ptr0[i0];
                auto tmp1 = in_ptr1[i0];
                auto tmp2 = (tmp1 != tmp1) ? tmp1 : std::min(tmp0, tmp1);
                tmp3 += tmp2;
            }
            out_ptr0[0] = tmp3;
        }
    }
}

extern "C" void kernel(const float* __restrict__ in_ptr0,
                       const float* __restrict__ in_ptr1,
                       float* __restrict__ out_ptr0)
{
    {
        {
            float tmp3 = 0;
            for(long i0=0; i0<96; i0+=1)
            {
                auto tmp0 = in_ptr0[i0];
                auto tmp1 = in_ptr1[i0];
                auto tmp2 = (tmp1 != tmp1) ? tmp1 : std::min(tmp0, tmp1);
                tmp3 += tmp2;
            }
            out_ptr0[0] = tmp3;
        }
    }
}
```

Now with `dynamic=True` as well:
```python
import torch
import time

min_sum = lambda a, b: torch.min(a, b).sum(-1)
min_sum_compiled = torch.compile(min_sum)

torch._dynamo.reset()
min_sum_compiled_dynamic = torch.compile(min_sum, dynamic = True)

static_shapes = [
    ((192,), (192,)),
    ((96,), (96,))
]

dynamic_shapes = [
    ((6, 183, 1, 192), (6, 1, 183, 192)),
    ((6, 183, 1, 96), (6, 1, 183, 96))
]

for k in range(100):
    for a_shape, b_shape in (static_shapes + dynamic_shapes):
        a = torch.rand(*a_shape, dtype = torch.float32)
        b = torch.rand(*b_shape, dtype = torch.float32)
        min_sum(a, b)
        min_sum_compiled(a, b)
        min_sum_compiled_dynamic(a, b)

for a_shape, b_shape in (static_shapes + dynamic_shapes):
    a = torch.rand(*a_shape, dtype = torch.float32)
    b = torch.rand(*b_shape, dtype = torch.float32)
    tic = time.time()
    for k in range(100):
        min_sum(a, b)
    print('min_sum', a_shape, b_shape, (time.time() - tic) * 1000, 'ms')

    tic = time.time()
    for k in range(100):
        min_sum_compiled(a, b)
    print('min_sum_compiled', a_shape, b_shape, (time.time() - tic) * 1000, 'ms')

    tic = time.time()
    for k in range(100):
        min_sum_compiled_dynamic(a, b)
    print('min_sum_compiled_dynamic', a_shape, b_shape, (time.time() - tic) * 1000, 'ms')
```

which gives:
```
/home/vadimkantorov/.local/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py:279: UserWarning: changing options to `torch.compile()` may require calling `torch._dynamo.reset()` to take effect
  warnings.warn(
/home/vadimkantorov/.local/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py:279: UserWarning: changing options to `torch.compile()` may require calling `torch._dynamo.reset()` to take effect
  warnings.warn(
/home/vadimkantorov/.local/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py:279: UserWarning: changing options to `torch.compile()` may require calling `torch._dynamo.reset()` to take effect
  warnings.warn(
/home/vadimkantorov/.local/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py:279: UserWarning: changing options to `torch.compile()` may require calling `torch._dynamo.reset()` to take effect
  warnings.warn(
min_sum (192,) (192,) 2.2292137145996094 ms
min_sum_compiled (192,) (192,) 6.571292877197266 ms
min_sum_compiled_dynamic (192,) (192,) 11.109352111816406 ms
min_sum (96,) (96,) 1.6977787017822266 ms
min_sum_compiled (96,) (96,) 5.49626350402832 ms
min_sum_compiled_dynamic (96,) (96,) 9.80377197265625 ms
min_sum (6, 183, 1, 192) (6, 1, 183, 192) 34465.22498130798 ms
min_sum_compiled (6, 183, 1, 192) (6, 1, 183, 192) 645.46799659729 ms
min_sum_compiled_dynamic (6, 183, 1, 192) (6, 1, 183, 192) 708.4212303161621 ms
min_sum (6, 183, 1, 96) (6, 1, 183, 96) 16976.492643356323 ms
min_sum_compiled (6, 183, 1, 96) (6, 1, 183, 96) 375.8273124694824 ms
min_sum_compiled_dynamic (6, 183, 1, 96) (6, 1, 183, 96) 364.3450736999512 ms
```

Note the warnings, despite actually calling `torch._dynamo.reset()`. It's also a weird design, as one assumes that args to previous torch.compile calls do not affect the current call. Also, why do I get this warning 4 times?

Logs: [torch_compile_debug.zip](https://github.com/pytorch/pytorch/files/12242643/torch_compile_debug.zip)

So we can see here that compiled version starts to be faster than the eager one.

Also, `dynamic = True` is slower than for 6x183x192 but faster for 6x183x96. But I could not find an `output_code.py` file for it though. Seems fishy!

Kernels that I did find:
```cpp
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       const float* __restrict__ in_ptr1,
                       float* __restrict__ out_ptr0)
{
    #pragma omp parallel num_threads(4)
    {
        {
            #pragma omp for  collapse(2)
            for(long i0=0; i0<6; i0+=1)
            {
                for(long i1=0; i1<183; i1+=1)
                {
                    #pragma GCC ivdep
                    for(long i2=0; i2<183; i2+=1)
                    {
                        {
                            float tmp3 = 0;
                            for(long i3=0; i3<96; i3+=1)
                            {
                                auto tmp0 = in_ptr0[i3 + (96*i1) + (17568*i0)];
                                auto tmp1 = in_ptr1[i3 + (96*i2) + (17568*i0)];
                                auto tmp2 = (tmp1 != tmp1) ? tmp1 : std::min(tmp0, tmp1);
                                tmp3 += tmp2;
                            }
                            out_ptr0[i2 + (183*i1) + (33489*i0)] = tmp3;
                        }
                    }
                }
            }
        }
    }
}

extern "C" void kernel(const float* __restrict__ in_ptr0,
                       const float* __restrict__ in_ptr1,
                       float* __restrict__ out_ptr0)
{
    #pragma omp parallel num_threads(4)
    {
        {
            #pragma omp for  collapse(2)
            for(long i0=0; i0<6; i0+=1)
            {
                for(long i1=0; i1<183; i1+=1)
                {
                    #pragma GCC ivdep
                    for(long i2=0; i2<183; i2+=1)
                    {
                        {
                            float tmp3 = 0;
                            for(long i3=0; i3<192; i3+=1)
                            {
                                auto tmp0 = in_ptr0[i3 + (192*i1) + (35136*i0)];
                                auto tmp1 = in_ptr1[i3 + (192*i2) + (35136*i0)];
                                auto tmp2 = (tmp1 != tmp1) ? tmp1 : std::min(tmp0, tmp1);
                                tmp3 += tmp2;
                            }
                            out_ptr0[i2 + (183*i1) + (33489*i0)] = tmp3;
                        }
                    }
                }
            }
        }
    }
}
```

Both kernels have burned in static shapes. I do not mind to have the variants for 96 and 192 (as these are embedding dim sizes, but I would like to have other dims dynamic, at least for the (BxNxC, BxMxC) -> BxNxM where B,N,M are dynamic)

At the end of the day I'm interested in GPU as well. 

### Versions
2.0.1+cpu

**Comment by vadimkantorov** [comment]
With forcing `ATEN_CPU_CAPABILITY=avx512`:

I get:
```
min_sum (192,) (192,) 1.154184341430664 ms
min_sum_compiled (192,) (192,) 3.445148468017578 ms
min_sum_compiled_dynamic (192,) (192,) 5.773305892944336 ms
min_sum (96,) (96,) 0.8625984191894531 ms
min_sum_compiled (96,) (96,) 2.580881118774414 ms
min_sum_compiled_dynamic (96,) (96,) 3.8628578186035156 ms
min_sum (6, 183, 1, 192) (6, 1, 183, 192) 16062.309265136719 ms
min_sum_compiled (6, 183, 1, 192) (6, 1, 183, 192) 260.2200508117676 ms
min_sum_compiled_dynamic (6, 183, 1, 192) (6, 1, 183, 192) 206.89797401428223 ms
min_sum (6, 183, 1, 96) (6, 1, 183, 96) 8093.605756759644 ms
min_sum_compiled (6, 183, 1, 96) (6, 1, 183, 96) 223.03295135498047 ms
min_sum_compiled_dynamic (6, 183, 1, 96) (6, 1, 183, 96) 159.4717502593994 ms
```

[log.txt](https://github.com/pytorch/pytorch/files/12243272/log.txt)
[torch_compile_debug.zip](https://github.com/pytorch/pytorch/files/12243273/torch_compile_debug.zip)

**Comment by vadimkantorov** [comment]
It seems that AVX512 seems helpful for perf, except 1d. Any ideas why?

It also seems that in actual `output_code.py` as opposed to `log.txt`, the dynamic versions were selected, not the static ones, including for 1d. Why is that?

I wanted to keep separate `min_sum_compiled` and `min_sum_compiled_dynamic`

**Comment by vadimkantorov** [comment]
Related issue about max-sum: https://github.com/pytorch/pytorch/issues/59545

**Comment by ezyang** [comment]
This is more of a cpu inductor question than it is a dynamic shapes question.

**Comment by vadimkantorov** [comment]
for dynamic shapes, that warning behavior is strange and why the option of one torch.compile actually affects the other torch.compile and selected kernels. it seems that after torch.compile(min_sum, dynamic = True) is called for the same function, it uses dynamic shapes even for the min_sum which was compile'd without dynamic shapes. And it's definitely hard to udnerstand from the logs what's going on :( quite big logs for such simple tasks.

The question on how to express which shapes do need to be dynamic and which shapes can be considered static is also a general one (similar to when we do torch.onnx.export we can specify manually the dynamic shapes).

The question about CPU capability discovery is also kind of general one, as indeed one gets AVX2 as default CPU capability (the CPU capability check is outside inductor, but still affects it?)

The question about logging: it seems that on nightly `TORCH_LOGS=output_code` works, but I think there should be a single docs webpage `all environment variables used by Torch` (maybe one page for runtime and one page for compiling from source)

**Comment by vadimkantorov** [comment]
Could you please also upload your test program?

Basically, I would like to have both static and dynamic shape versions in the same program: static for 1d version; and semi-dynamic-static for the 2d matmul-like version (`semi` - because I would like to keep the leading dimension dynamic, and the inner dimension static).

I had signs that single usage of `min_sum_compiled_dynamic` (dynamic=True) affected the usage and kernels of the static `min_sum_compiled`.

How can I check what assembly got produced? I'm interested whether the static shape versions got unrolled.

Ideally, I would also like to have an option to ask to remove nan handling (I know for sure that I would not have nans as inputs)

**Comment by jgong5** [comment]
> Could you please also upload your test program?

```python
# foo.py
import torch
import time

# parse command line args
import argparse
parser = argparse.ArgumentParser()
parser.add_argument('--dynamic', action='store_true')
parser.add_argument('--simdlen', type=int, default=512)
args = parser.parse_args()

from torch._inductor import config
config.cpp.simdlen = args.simdlen
min_sum = lambda a, b: torch.min(a, b).sum(-1)
min_sum_compiled = torch.compile(min_sum, dynamic=args.dynamic)

static_shapes = [
    ((192,), (192,)),
    ((96,), (96,))
]

# warmup
for k in range(100):
    for a_shape, b_shape in static_shapes:
        a = torch.rand(*a_shape, dtype = torch.float32)
        b = torch.rand(*b_shape, dtype = torch.float32)
        min_sum(a, b)
        min_sum_compiled(a, b)

for a_shape, b_shape in static_shapes:
    a = torch.rand(*a_shape, dtype = torch.float32)
    b = torch.rand(*b_shape, dtype = torch.float32)
    tic = time.time()
    for k in range(100):
        min_sum(a, b)
    print('min_sum', a_shape, b_shape, (time.time() - tic) * 1000, 'ms')

    tic = time.time()
    for k in range(100):
        min_sum(a, b)
    print('min_sum_compiled', a_shape, b_shape, (time.time() - tic) * 1000, 'ms')
```

> How can I check what assembly got produced? I'm interested whether the static shape versions got unrolled (there is no pragma unroll for the static version).

Yes, it is unrolled. You may add a print for the `output_path` in `CppCodeCache.load` and then objdump it.

> Ideally, I would also like to have an option to ask to remove nan handling (I know for sure that I would not have nans as inputs)

Does it matter a lot? It is only in the tail loop. What is the perf gain without it?

**Comment by vadimkantorov** [comment]
How can I do `CppCodeCache.load`? https://pytorch.org/docs search return nothing :( Thanks :) 

Regarding nan handling - I would assume the drop is noticeable for small 1d vectors, for larger inputs probably memory loads/stores will dominate

**Comment by vadimkantorov** [comment]
The follow-up issue is : https://github.com/pytorch/pytorch/issues/106614

**Comment by jgong5** [comment]
> How can I do `CppCodeCache.load`? https://pytorch.org/docs search return nothing :( Thanks :)

I was referring to this line: https://github.com/pytorch/pytorch/blob/416bf4e3e7c5f90bbc306e55501881b4f8bde981/torch/_inductor/codecache.py#L1035

You can dump the `output_path` to get the dynamic library file.
